### PR TITLE
Postfix stats - Rewrited plugin

### PR DESCRIPTION
**Pull request by CristianDeluxe**
Hi, i've followed all your suggestions from previous pull request, unfortunately i can't remove merges and finally i've created another pull request 😐 

Previous pull:
https://github.com/munin-monitoring/contrib/pull/711

Plugin has been completely rewritten:

- No use of TMP files 👍 
- Much faster because "pflogsumm" command has been tuned
- More efficient "parsing" now it use grep and regexp, and is more "failsafe"
- Added autoconf
- Script more documented
- Removed duplicated code as far as possible
- Removed backticks

I hope everything is fine!

### CODE REVIEW

File: plugins/mail/postfix_stats [hunk 1]
```diff
@@ -3,50 +3,102 @@
...
+# Autoconf Section
+#
+if [ "$1" = 'autoconf' ]; then
+	# Try to find pflogsumm with default name
+	PFLOG_EXIST=$(command -v pflogsumm.pl)
+
+	# Try to find pflogsumm without any extension
+	if [[ -z "${PFLOG_EXIST}" ]]
+		then
+		PFLOG_EXIST=$(command -v pflogsumm)
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: You are using path name guessing for determing the location of "pflogsum". This is a good approach.
But you do not use the result of this guessing below in the actual "fetch" branch.

I would propose the following approach:
```
PFLOGSUM="${pflogsum:-}"
[ -z "$PFLOGSUM" ] && PFLOGSUM="$(which pflogsum pflogsum.pl | head -1)"
```

File: plugins/mail/postfix_stats [hunk 2]
```diff
@@ -3,50 +3,102 @@
...
+	fi
+
+	if [[ -z "${PFLOG_EXIST}" ]]
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: Style hint: finishing the `if` line with `; then` is quite common.

File: plugins/mail/postfix_stats [hunk 3]
```diff
@@ -3,50 +3,102 @@
...
+		then
+		echo 'no';
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: you can save one character here :)

Additionally you should add the reason for "no" in brackets afterwards. E.g.: `echo "no (failed to find pflogsum)"`

File: plugins/mail/postfix_stats [hunk 4]
```diff
@@ -3,50 +3,102 @@
...
+	else
+		echo 'yes'
+	fi
+	exit 0;
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: same as above

File: plugins/mail/postfix_stats [hunk 5]
```diff
@@ -3,50 +3,102 @@
...
+if [ "$1" = 'config' ]; then
+	echo 'graph_title Postfix statistics'
+	echo 'graph_vlabel Postfix statistics'
+	echo 'graph_category mail'
+	echo 'graph_scale no'
+	echo 'graph_period minute'
+	echo 'graph_total Total'
+
+	# Generate config for each field
+	for i in "${FIELDS_ARR[@]}"
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: if you moved away from `declare` above, then you can use `for i in $FIELDS_ARR; do`

File: plugins/mail/postfix_stats [hunk 6]
```diff
@@ -3,50 +3,102 @@
...
+TMP_RAW=$("${PFLOGSUMM}" -d today --detail 0 --zero-fill "${SYS_LOG}" "${SYS_LOG2}")
+
+# Parse value from Raw result
+# 
+# Return digit if regex are parsed correctly
+#
+# Return -1 if any error occurs
+#
+parseValue() {
+	TMP_RETURN=$(echo "${TMP_RAW}" | grep -Ei '^[[:space:]]+[[:digit:]]+[[:space:]]+'"${1}"'.*$' | grep -oEi '[[:digit:]]+[[:space:]]+' | head -n 1 | sed 's: ::g')
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: You should use `local` for variables in functions.
Style hint: wrapping the first `grep` expression in double-quotes should be more readable.

File: plugins/mail/postfix_stats [hunk 7]
```diff
@@ -3,50 +3,102 @@
...
+	if [[ -z "${TMP_RETURN}" ]]
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: bashism `[[ .. ]]`

File: plugins/mail/postfix_stats [hunk 8]
```diff
@@ -3,50 +3,102 @@
...
+		then
+		echo -1
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: probably you want to use `U` ("undefined") instead

File: plugins/mail/postfix_stats [hunk 9]
```diff
@@ -3,50 +3,102 @@
 # Plugin to show Postfix statistics - needs pflogsumm
```
Comment (sumpfralle) [inline plugins/mail/postfix_stats]: If you like perfection, you may want to transform the current plugin documentation into the POD format (useful for `munindoc postfix_stats`). See the plugin "processes/multimemory" for an example.

**Comment by CristianDeluxe** [comment]
Thanks for your suggestions and tips 👍  now the script is strictly "SH" and better than before.

I'm not a sh/bash programmer, usually i use it for do some works in my servers, but with this i've learned a lot of stuff!

Thanks again!
Chris

**Comment by sumpfralle** [comment]
Hi Chris,
I am glad, that you enjoy the progress! :)

I have one last (really!) nitpick:
you use `eval` in the pflogsumm call - probably to work around shellcheck's complain about the missing quotes around SYS_LOG (due to the space separated filenames). `eval` works here, but since it can have a variety of side-effects, you should try to avoid it - your fellow readers will appreciate this.
Just use your original line, remove the quotes around SYS_LOG and add a `shellcheck disable=???` comment (see its manpage) in order to silence shellcheck's warning (since you know what you are doing and there is no better way).

**Comment by CristianDeluxe** [comment]
Here we go again! 😄 

All the changes made! Thank you for being so conscientious!

When I have some free time I will continue working on other plugins that can be polished a little more

It has been a pleasure working with you!

regards

**Comment by sumpfralle** [comment]
thank you for your kind words!
Polishing plugins is really welcome!

**Pull request merged**